src/irr/asset/COBJMeshFileLoader.cpp

```diff
@@ -166,101 +182,80 @@ asset::SAssetBundle COBJMeshFileLoader::loadAsset(io::IReadFile* _file, const as
 			break;
 
 		case 'g': // group name
-			{
-				char grp[WORD_BUFFER_LENGTH];
-				bufPtr = goAndCopyNextWord(grp, bufPtr, WORD_BUFFER_LENGTH, bufEnd);
-#ifdef _IRR_DEBUG_OBJ_LOADER_
-	os::Printer::log("Loaded group start",grp, ELL_DEBUG);
-#endif
-				if (ctx.useGroups)
-				{
-					if (0 != grp[0])
-						grpName = grp;
-					else
-						grpName = "default";
-
-                    asset::IAsset::E_TYPE types[] {asset::IAsset::ET_SUB_MESH, (asset::IAsset::E_TYPE)0u };
-                    auto mb_bundle = _override->findCachedAsset(genKeyForMeshBuf(ctx, _file->getFileName().c_str(), mtlName, grpName), types, ctx.inner, 1u).getContents();
-                    if (mb_bundle.first!=mb_bundle.second)
-                    {
-                        auto mb = static_cast<asset::ICPUMeshBuffer*>(mb_bundle.first->get());
-                        mb->grab();
-                        SObjMtl* mtl = findMtl(ctx, mtlName, grpName);
-                        ctx.preloadedSubmeshes.insert(std::make_pair(mtl, mb));
-                    }
-                    else mtlChanged=true;
-
-                    submeshLoadedFromCache = (mb_bundle.first!=mb_bundle.second);
-				}
-			}
+            bufPtr = goNextWord(bufPtr, bufEnd);
 			break;
-
 		case 's': // smoothing can be a group or off (equiv. to 0)
 			{
-				char smooth[WORD_BUFFER_LENGTH];
-				bufPtr = goAndCopyNextWord(smooth, bufPtr, WORD_BUFFER_LENGTH, bufEnd);
+				bufPtr = goAndCopyNextWord(tmpbuf, bufPtr, WORD_BUFFER_LENGTH, bufEnd);
 #ifdef _IRR_DEBUG_OBJ_LOADER_
-	os::Printer::log("Loaded smoothing group start",smooth, ELL_DEBUG);
+	os::Printer::log("Loaded smoothing group start",tmpbuf, ELL_DEBUG);
 #endif
-				if (core::stringc("off")==smooth)
-					smoothingGroup=0;
+				if (core::stringc("off")==tmpbuf)
+					smoothingGroup=0u;
 				else
-                    sscanf(smooth,"%u",&smoothingGroup);
+                    sscanf(tmpbuf,"%u",&smoothingGroup);
 			}
 			break;
 
 		case 'u': // usemtl
 			// get name of material
 			{
-				char matName[WORD_BUFFER_LENGTH];
-				bufPtr = goAndCopyNextWord(matName, bufPtr, WORD_BUFFER_LENGTH, bufEnd);
+				bufPtr = goAndCopyNextWord(tmpbuf, bufPtr, WORD_BUFFER_LENGTH, bufEnd);
 #ifdef _IRR_DEBUG_OBJ_LOADER_
-	os::Printer::log("Loaded material start",matName, ELL_DEBUG);
+	os::Printer::log("Loaded material start",tmpbuf, ELL_DEBUG);
 #endif
-				mtlName=matName;
+				mtlName=tmpbuf;
 
                 if (ctx.useMaterials && !ctx.useGroups)
                 {
                     asset::IAsset::E_TYPE types[] {asset::IAsset::ET_SUB_MESH, (asset::IAsset::E_TYPE)0u };
                     auto mb_bundle = _override->findCachedAsset(genKeyForMeshBuf(ctx, _file->getFileName().c_str(), mtlName, grpName), types, ctx.inner, 1u).getContents();
```

Comment:
from now on when using `findCachedAsset` or similar functions in a loader that take a hierarchy level, abandon the constants.

as hierarchy level use `_hierarchyLevel+ICPUMeshBuffer::PIPELINE_HIERARCHYLEVELS_BELOW`

obviously you need to define these magical `_HIERARCHYLEVELS_BELOW` constexpr constants for all the asset classes.
https://github.com/buildaworldnet/IrrlichtBAW/blob/shader_pipeline/include/irr/asset/ICPUMesh.h#L18

P.S. This is actually a thing to refactor in all the Mesh Loaders

Reply:
done in OBJ loader